### PR TITLE
eTkt 85 autocomplete in begin and end

### PR DESCRIPTION
**Pull request by Mikhail-M**

### CODE REVIEW

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 1]
```diff
@@ -95,13 +118,24 @@ define((require, exports, module) ->
...
           "storage.type"
+          "lparen." + ENVIRONMENT_TOKENTYPE
+          "variable.parameter." + ENVIRONMENT_TOKENTYPE
+          "rparen"
           "lparen"
           "variable.parameter"
           "rparen"
         ]
-        regex: "(\\\\(?:begin|end))({)(\\w*)(})"
+        regex: "(\\\\(?:begin|end))({)(tabular)(})({)(.*)(})"
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: What `tabular` is doing here?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 2]
```diff
@@ -95,13 +118,24 @@ define((require, exports, module) ->
...
+      }
+      {
+        token: [
+          "storage.type"
+          "lparen.environment." + ENVIRONMENT_TOKENTYPE
+          "variable.parameter." + ENVIRONMENT_TOKENTYPE
+          "rparen"
+        ]
+        regex: "(\\\\(?:begin|end))({)(.*)(})"
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: out of curiosity: are all these capture groups in the regexp used? Can we remove parentheses around `{` and `}`?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 1 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: Nothing
Forget it, please

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 2 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: No, we can't
Backend need to () for every token

File: lib/ace/autocomplete/text_completer.js [hunk 3]
```diff
@@ -3,7 +3,7 @@
  *
  * Copyright (c) 2012, Ajax.org B.V.
  * All rights reserved.
- * 
+ *
```
Comment (dbarashev) [inline lib/ace/autocomplete/text_completer.js]: pls revert changes in this file (they seem to be whitespace-only)

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 4]
```diff
@@ -2,11 +2,23 @@ foo = null # ACE builder wants some meaningful JS code here to use ace.define in
...
+
+    EQUATION_TOKENTYPE = PapeeriaLatexHighlightRules.EQUATION_TOKENTYPE
+    LIST_TOKENTYPE = PapeeriaLatexHighlightRules.LIST_TOKENTYPE
+    ENVIRONMENT_TOKENTYPE = PapeeriaLatexHighlightRules.ENVIRONMENT_TOKENTYPE
+    FIGURE_TOKENTYPE = PapeeriaLatexHighlightRules.FIGURE_TOKENTYPE
+    TABLE_TOKENTYPE = PapeeriaLatexHighlightRules.TABLE_TOKENTYPE
 
-    EQUATION_TOKENTYPE  = PapeeriaLatexHighlightRules.EQUATION_TOKENTYPE
-    LIST_TOKENTYPE  = PapeeriaLatexHighlightRules.LIST_TOKENTYPE
+    #order if importance
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: what did you want to say by this comment? (nit: space after `#` pls)

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 5]
```diff
@@ -6,15 +6,38 @@ define((require, exports, module) ->
...
   LIST_REGEX = "itemize|enumerate"
   EQUATION_REGEX = "equation|equation\\*"
+
+  TABLE_REGEX = "table"
+  FIGURE_REGEX = "figure"
+
   LIST_STATE = "list"
   EQUATION_STATE = "equation"
+  ENVIRONMENT_STATE = "environment"
+  TABLE_STATE = 'table'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: pls use the same quotes for strings. Either double or single

File: lib/ace/snippets.js [hunk 6]
```diff
@@ -46,7 +46,7 @@ var SnippetManager = function() {
 
 (function() {
     oop.implement(this, EventEmitter);
-    
+
```
Comment (dbarashev) [inline lib/ace/snippets.js]: pls keep the whitespace in this file. We don't need merge issues out of nowhere

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 5 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: it is my curse

File: lib/ace/snippets.js [hunk 7]
```diff
@@ -960,3 +960,4 @@ var Editor = require("./editor").Editor;
 }).call(Editor.prototype);
 
 });
+
```
Comment (dbarashev) [inline lib/ace/snippets.js]: what did you mean to say with this commit?

**Pull request merged**